bindings/c/src/error.rs

```diff
@@ -0,0 +1,203 @@
+///         [`od::ErrorKind`]
+#[repr(C)]
+pub enum opendal_errno {
+    /// returning it back. For example, s3 returns an internal service error.
+    Unexpected,
+    /// Underlying service doesn't support this operation.
+    Unsupported,
+    /// The config for backend is invalid.
+    ConfigInvalid,
+    /// The given path is not found.
+    NotFound,
+    /// The given path doesn't have enough permission for this operation
+    PermissionDenied,
+    /// The given path is a directory.
+    IsADirectory,
+    /// The given path is not a directory.
+    NotADirectory,
+    /// The given path already exists thus we failed to the specified operation on it.
+    AlreadyExists,
+    /// Requests that sent to this path is over the limit, please slow down.
+    RateLimited,
+    /// The given file paths are same.
+    IsSameFile,
+    /// Unknown error, since [`opendal::ErrorKind`] is nonexhaustive
+    UnknownError,
+    /// NullPtr error, meaning that a provided pointer which need to be dereferenced is null
+    NullPtr,
+}
+
+/// The C-compatible error status for opendal::Error.
+/// NOTICE: the type definition of [`ErrorNo`] should be aligned with
+///         [`od::ErrorStatus`]
+#[repr(C)]
+pub enum opendal_error_status {
+    /// Permanent means without external changes, the error never changes.
+    ///
+    /// For example, underlying services returns a not found error.
+    ///
+    /// Users SHOULD never retry this operation.
+    Permanent,
+    /// Temporary means this error is returned for temporary.
+    ///
+    /// For example, underlying services is rate limited or unavailable for temporary.
+    ///
+    /// Users CAN retry the operation to resolve it.
+    Temporary,
+    /// Persistent means this error used to be temporary but still failed after retry.
+    ///
+    /// For example, underlying services kept returning network errors.
+    ///
+    /// Users MAY retry this operation but it's highly possible to error again.
+    Persistent,
+}
+
+impl OpendalError {
+    /// Convert the [`od::ErrorKind`] of [`od::Error`] to our own
+    /// C-compatible type
+    pub(crate) fn errno(&self) -> opendal_errno {
+        let e = &self.0;
+        match e.kind() {
+            od::ErrorKind::Unexpected => opendal_errno::Unexpected,
+            od::ErrorKind::Unsupported => opendal_errno::Unsupported,
+            od::ErrorKind::ConfigInvalid => opendal_errno::ConfigInvalid,
+            od::ErrorKind::NotFound => opendal_errno::NotFound,
+            od::ErrorKind::PermissionDenied => opendal_errno::PermissionDenied,
+            od::ErrorKind::IsADirectory => opendal_errno::IsADirectory,
+            od::ErrorKind::NotADirectory => opendal_errno::NotADirectory,
+            od::ErrorKind::AlreadyExists => opendal_errno::AlreadyExists,
+            od::ErrorKind::RateLimited => opendal_errno::RateLimited,
+            od::ErrorKind::IsSameFile => opendal_errno::IsSameFile,
+            _ => opendal_errno::UnknownError,
```

Comment:
I don't think we need unknown_error.

Reply:
If we are using pattern matching, we either panics by `unreachable!()` or add a placeholder like this, what's your opinion?

Reply:
> If we are using pattern matching, we either panics by `unreachable!()` or add a placeholder like this, what's your opinion?

All bindings must keep sync with `core`. So we will meet compile failed here.